gossip/sfc_index.go

```diff
@@ -271,20 +271,39 @@ func (s *Service) processSfc(block *inter.Block, receipts types.Receipts, blockF
 	// Write cheaters
 	for _, stakerID := range cheaters {
 		staker := s.store.GetSfcStaker(stakerID)
-		if !staker.IsCheater {
-			// write into node DB
-			staker.IsCheater = true
-			s.store.SetSfcStaker(stakerID, staker)
-			// write into SFC contract
-			position := sfcpos.Staker(stakerID)
-			statedb.SetState(sfc.ContractAddress, position.IsCheater(), common.BytesToHash([]byte{1}))
+		if staker.HasFork() {
+			continue
 		}
+		// write into DB
+		staker.Status |= sfctype.FORK_BIT
+		s.store.SetSfcStaker(stakerID, staker)
+		// write into SFC contract
+		position := sfcpos.Staker(stakerID)
+		statedb.SetState(sfc.ContractAddress, position.Status(), utils.U64to256(staker.Status))
 	}
 
 	if sealEpoch {
 		epoch256 := utils.U64to256(uint64(epoch))
 		statedb.SetState(sfc.ContractAddress, sfcpos.CurrentSealedEpoch(), epoch256)
 
+		// Write offline validators
+		for _, it := range s.store.GetSfcStakers() {
+			if it.Staker.Offline() {
+				continue
+			}
+
+			gotMissed := s.store.GetBlocksMissed(it.StakerID)
+			badMissed := s.config.Net.Economy.OfflinePenaltyThreshold
+			if gotMissed.Num >= badMissed.Num && gotMissed.Period >= inter.Timestamp(badMissed.Period) {
```

Comment:
add some logs?
what if the condition at 297 does not hold? ignore writes? please add a comment